### PR TITLE
Remove `waitDuration` from skill representation and use arrays instead of linked lists

### PR DESCRIPTION
**Pull request by zqsz-xiv**
Part 1 of invalid timeline support. This PR removes `waitDuration` from all non-wait actions, and uses arrays to represent `Line` objects in place of clunky linked lists.

~~**This is currently a breaking change for users** (my own FRU timeline actually is broken because of this). I still need to think about how best to avoid breaking user timelines, and I'll probably end up needing to add an extra pass that detects if a skill node has a non-standard `waitDuration` and modify/generate additional "wait" nodes accordingly. Alternatively, we could perform a shellVersion bump and just have two different replay behaviors, but that sounds annoying to maintain.~~

**Edit:** This is still a breaking change, but existing timelines (including those created in 7.0) will import properly, though some saved waitDuration values in explicit "wait" events may be changed. My own FRU timeline now imports fine, and in fact I was able to import my 7.05 M4S BLM timeline that got broken due to Despair becoming instant-cast.

TODO:
- ~~Remove the "since [now | last action]" dropdown, since it always implicitly is "now." I honestly didn't realize this dropdown existed until pre-reviewing the PR myself.~~ The branch now works with this dropdown.
- Fix BRD and WAR tests, which are failing because they make assumptions about time advances in the test controller that are now invalid.
- Fix BLM line load test, which is broken because the cursor no longer moves past the final paradox's damage event.

New behavior changes, help text + changelog still needs to be edited to communicate these to users:
- No longer need to press "space" to advance in "manual" mode
- No longer need to double click to perform an action blocked by cd/animation
- Wait events now **always** calculate from the end of the previous animation lock. In certain scenarios (notably single-clicking a button to wait for the animation to complete), explicit wait events may be generated to indicate that the cursor should move, and will not be "straightened'" by timeline editor actions.

I've tested most of the following behaviors already while developing, but I'll go through this checklist again later to make sure I didn't accidentally change anything. Some of these could also be added to automated tests. If anyone can think of edge cases that these aren't covered, please lmk and I'll add it!

- [x] Add skills + waits in real-time tick mode
- [x] Add skills + waits in manual tick mode
- [x] Click on arbitrary point in timeline between already-used skills, and check buffs
- [x] Check dot uptime % + tick count in tables
- [x] Select single skill via timeline canvas, view damage statistics + gauge state
- [x] ^ ... multiple skills ...
- [x] ... single skill via timeline editor ..., and move it around
- [x] ... multiple skills via timeline editor ..., and move it around
- [x] ... multiple actions via timeline editor, starting/ending with a wait
- [x] Use a GCD when resources are unavailable (e.g. F4 at 0 MP), then use an ability after
- [x] Use a skill w/ cooldown, then attempt to use the skill w/ cooldown again so the cursor advances but resources are not available when the use attempt is made (e.g. amplifier); then use another skill
- [x] Moving wait actions around
- [x] Moving skills across wait actions
- [x] Perform multiple sequential timeline edits
- [x] Adding a wait with "since last action" selected, and doing so after selecting other timelines + historical skills
- [x] Save/load new presets with waits
- [x] Save/load full timelines with waits
- [x] Swap the selected timeline and ensure no changes happened due to inconsistencies with replay modes
- [x] Load existing presets
- [x] Load existing timeline created in "manual" mode w/ wait relative to START of previous skill usage (this will probably be changed, but not sure if there's anything we can do about it)
- [x] Load existing timelines containing buff toggles
- [x] Load test corpus
- [x] Load timelines that contain a wait as the final action
- [x] Load timelines that have implicit waits at the end of a GCD as the final action
- [x] Perform skill that causes an interrupted cast
- [x] Load a timeline that causes an interrupted cast
- [x] Rearrange a timeline s.t. a cast is interrupted

Some minor perf improvements, demonstrated by loading Akairyu's 2286 sps top file, which is 19 min long:
- Loading the page with 4 timelines in localstorage: about 0.57s to paint in both cases
- Importing a timeline: 324ms -> 160ms

### CODE REVIEW

**Review by miyehn — COMMENTED** [review]
From a glance I think most changes look good. One thing that concerns me a little is using `index` or `actionIndex` to refer to actions (nodes).
* In some places `index` and `node` are both provided as function arguments, and they essentially refer to the same thing which is a bit repetitive? And if they are ever out of sync it can be very confusing :0
* In some other places only `index` is provided as a function argument. But what it refers to depends on the context (which line/record is currently active, etc.) and may change over time. For example, is there any special handling needed when the player reorders the timeline? What if the player deletes some actions and makes some indices invalid? What about switching active slot? I messed around a bit and it seems fine so far. Maybe there is already correctness guarantee. It just seems less obvious I think :0
  * not that comparing `ActionNode` pointers and/or global node id solves all these problems, but at least they are... more clear?

As for performance, any optimization is nice but I think development time and speed is so valuable that as long as user experience is not so bad that people actively avoid features, we can usually just go with whatever's easier to read/maintain (aka if performance was a consideration for using indices over node pointers)

I didn't scrutinize every line though. And ultimately it's you maintaining the changes so however you think is easier👌

File: src/Controller/Record.ts [hunk 1]
```diff
@@ -1,111 +1,227 @@
...
-	skillName?: ActionKey;
-	buffName?: string;
+	// TODO split into different properties for dots/hots?
 	applicationTime?: number;
 	#dotOverrideAmount: Map<ResourceKey, number>;
 	#dotTimeGap: Map<ResourceKey, number>;
-	targetCount: number = 1;
 	#hotOverrideAmount: Map<ResourceKey, number>;
 	#hotTimeGap: Map<ResourceKey, number>;
 	healTargetCount: number = 1;
```
[the remark this answers is not in the document]
Comment (zqsz-xiv) [reply src/Controller/Record.ts:142]: Good point, I'll move it over.

**Comment by zqsz-xiv** [comment]
Regarding passing index vs. object pointer: perf improvement wasn't the main goal (the numbers are already well within human-noticeable latency anyway), but I found it a lot easier to reason about for basically two reasons:
1. It's a lot easier to manipulate an array than it is a linked list, and there are a lot of places in the current codebase that have to manually dip across an abstraction barrier to edit the `ActionNode` linked list. Technically this could be resolved by more aggressively using methods on the `Line` class to modify/edit the underlying list, but it's still kind of brittle and it takes a lot of effort to ensure the linked list;s invariants are maintained.
2. In any situation where an object has knowledge of the full `ActionNode[]`, it can get the corresponding index just by enumerating the array.

For example, look at the difference in complexity for `Record.moveSelected(offset)`, which moves the selected skills by `offset` positions in the timeline.

Old:
https://github.com/xivintheshell/xivintheshell/blob/8aca88893f57bb26f73cc56b1f389f60a02f184b/src/Controller/Record.ts#L525-L590

New:
https://github.com/xivintheshell/xivintheshell/blob/0eef339b1312779cf45ff74f9a5316a2a8179ed1/src/Controller/Record.ts#L681-L707

> In some places index and node are both provided as function arguments, and they essentially refer to the same thing which is a bit repetitive? And if they are ever out of sync it can be very confusing :0

I think in the two situations where both arguments are passed, the node and index are tied pretty tightly together. Strictly speaking we could get away with passing some properties of the node instead of the whole object, but I found it easier to just keep it closer to the original code.
1. In `Controller.reportInterruption`, we need to format the name of the errored node, while also tracking the index to rewind simulation back to. The `node` argument is just to get the action's name.
2. In `TimelineActionElement`, we need to render information about the node, and keep the index to report selection positions back to the controller. Each `TimelineActionElement` is generated in a loop, so it's safe to assume the index/node pair is always valid.

> In some other places only index is provided as a function argument. But what it refers to depends on the context (which line/record is currently active, etc.) and may change over time.

I do think this is a fair point, but the only place I really found this to be a pain was when adding the `Controller.lastSkillTime` field to track when the last skill was used since I had to make sure it wasn't being updated by historical replays. In general though, I think the validity of an index is guaranteed by how the index is derived directly from the array of nodes. Most places where we previously passed around a node object were for the purpose of "iterate until this node is reached" or "iterate until one before this node," and I think it's just much easier to reason about + maintain correctness with a numeric index than a chain of node objects.

**Comment by miyehn** [comment]
Gotcha, I was actually thinking about using `ActionNode` array + consider passing nodes around instead of their indices, but yup as long as you know indices work well 👌 👌

**Comment by zqsz-xiv** [comment]
BTW @miyehn there's some text changes in `PlaybackControl.tsx` + `IntroSection.tsx`. Feel free to push changes yourself if there's anything off.
